src/storage/src/source/reclock/compat.rs

```diff
@@ -400,22 +433,53 @@ impl RemapHandle for PersistHandle {
     }
 
     async fn compact(&mut self, new_since: Antichain<Self::IntoTime>) {
-        if !PartialOrder::less_equal(self.read_handle.since(), &new_since) {
+        if !PartialOrder::less_equal(self.since_handle.since(), &new_since) {
             panic!(
                 "ReclockFollower: `new_since` ({:?}) is not beyond \
                 `self.since` ({:?}).",
                 new_since,
-                self.read_handle.since(),
+                self.since_handle.since(),
             );
         }
-        self.read_handle.maybe_downgrade_since(&new_since).await;
+        let epoch = self.since_handle.opaque().clone();
+        let result = self
+            .since_handle
+            .maybe_compare_and_downgrade_since(&epoch, (&epoch, &new_since))
+            .await;
+
+        if let Some(result) = result {
+            match result {
+                Ok(_) => {
+                    // All's well!
+                }
+                Err(current_epoch) => {
+                    // TODO(aljoscha): In the future, we might want to be
```

Comment:
I'd leave this in for later. wdyt?

Reply:
I think this works as indented and we don't want to do anything other than halting here. The controller would do exactly the same thing and this code logically belongs there so all is well